### PR TITLE
Migration tool for updating the encryption of secondary user store

### PR DESCRIPTION
**Pull request by inthirakumaaran**
## Purpose

Issue [wso2/product-is#4459](https://github.com/wso2/product-is/issues/4459)

After switching to internal keystore for encrypting secondary userstore password, this tool will help to decrypt all the existing secondary userstore password using the old keystore and re-encrypt them using the internal keystore

## Migrations (if applicable)
1. Build the project and take the org.wso2.is.password.migrate-1.0.jar under the org.wso2.is.password.migrate component
2. Put that jar into the dropins folder
3. Restart the wso2 server with -DpassMigrate option
4. After the migration completes, stop the server and remove added jar

### CODE REVIEW

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+     * @return if cipher text is not a cipher with meta data
+     */
+    public CipherHolder cipherTextToCipherHolder(byte[] cipherText) {
+
+        String cipherStr = new String(cipherText, Charset.defaultCharset());
+        try {
+            return gson.fromJson(cipherStr, CipherHolder.class);
+        } catch (JsonSyntaxException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Deserialization failed since cipher string is not representing cipher with metadata");
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java]: why is this logged in debug? is it expected?

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+                decryptedValue = "".getBytes();
+                if (log.isDebugEnabled()) {
+                    log.debug("Empty value for plainTextBytes null will persist to DB");
+                }
+            } else {
+                decryptedValue = keyStoreCipher.doFinal(cipherTextBytes);
+            }
+
+        } catch (Exception e) {
+            throw new CryptoException("errorDuringDecryption", e);
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java]: shall we add more context information here?

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+            } else {
+                keyStoreCipher = Cipher.getInstance("RSA", "BC");
+            }
+
+            keyStoreCipher.init(Cipher.DECRYPT_MODE, privateKey);
+
+            if (cipherTextBytes.length == 0) {
+                decryptedValue = "".getBytes();
+                if (log.isDebugEnabled()) {
+                    log.debug("Empty value for plainTextBytes null will persist to DB");
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java]: this is supposed to be secondary userstore password. so there is nothing to persist in the database.

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+                    ISMigrationServiceDataHolder.getServerConfigurationService(),
+                    ISMigrationServiceDataHolder.getRegistryService());
+            keyStore = keyMan.getPrimaryKeyStore();
+            privateKey = (PrivateKey) keyStore.getKey(primaryKeyStoreAlias, primaryKeyStoreKeyPass.toCharArray());
+            if (cipherTransformation != null) {
+                CipherHolder cipherHolder = cipherTextToCipherHolder(cipherTextBytes);
+                if (cipherHolder != null) {
+                    // cipher with meta data.
+                    if (log.isDebugEnabled()) {
+                        log.debug("Cipher transformation for decryption : " + cipherHolder.getTransformation());
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java]: more context information?

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/internal/ISMigrationServiceComponent.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    @Reference(
+            name = "registry.service",
+            service = RegistryService.class,
+            cardinality = ReferenceCardinality.MANDATORY,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unsetRegistryService"
+    )
+    protected void setRegistryService(RegistryService registryService) {
+
+        ISMigrationServiceDataHolder.setRegistryService(registryService);
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/internal/ISMigrationServiceComponent.java]: do we need all of these service references?

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/internal/ISMigrationServiceComponent.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    protected void activate(ComponentContext context) {
+
+        try {
+            String migration = System.getProperty(JVM_PROPERTY_MIGRATE_PASSWORD);
+            log.info("secondary userstore password migration jar added");
+
+            if (migration != null) {
+                UserStorePasswordMigrator userStorePasswordMigrator = new UserStorePasswordMigrator();
+                log.info("secondary userstore password migration started");
+                context.getBundleContext().registerService(UserStorePasswordMigrator.class, userStorePasswordMigrator,
```
Comment (madurangasiriwardena) [inline components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/internal/ISMigrationServiceComponent.java]: do we to register the migrator as a service?

Comment (inthirakumaaran) [reply components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/internal/ISMigrationServiceComponent.java]: No we can remove that

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/internal/ISMigrationServiceComponent.java (hunk 5 above)
Comment (inthirakumaaran) [reply components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/internal/ISMigrationServiceComponent.java]: No need

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java (hunk 1 above)
Comment (inthirakumaaran) [reply components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java]: we can shift this to info

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java (hunk 2 above)
Comment (inthirakumaaran) [reply components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java]: Sure

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java (hunk 3 above)
Comment (inthirakumaaran) [reply components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java]: Yes we can remove this

File: components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java (hunk 4 above)
Comment (inthirakumaaran) [reply components/org.wso2.is.password.migrate/src/main/java/org/wso2/is/password/secondaryUserStore/util/SecondaryUserstoreCryptoUtil.java]: added

**Pull request closed without merging**